server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java

```diff
@@ -8,42 +8,30 @@
 
 package org.elasticsearch.index.mapper;
 
+import java.util.Stack;
+
 public final class ContentPath {
 
     private static final char DELIMITER = '.';
 
     private final StringBuilder sb;
-
-    private int index = 0;
-
-    private String[] path = new String[10];
-
+    private final Stack<Integer> delimiterIndexes;
     private boolean withinLeafObject = false;
 
     public ContentPath() {
         this.sb = new StringBuilder();
-    }
-
-    String[] getPath() {
-        // used for testing
-        return path;
+        this.delimiterIndexes = new Stack<>();
     }
 
     public void add(String name) {
-        path[index++] = name;
-        if (index == path.length) { // expand if needed
-            expand();
-        }
-    }
-
-    private void expand() {
-        String[] newPath = new String[path.length + 10];
-        System.arraycopy(path, 0, newPath, 0, path.length);
-        path = newPath;
+        // Store the location of the previous final delimiter onto the stack,
+        // which will be the index of the 2nd last delimiter after appending the new name
+        delimiterIndexes.add(sb.length() - 1);
+        sb.append(name).append(DELIMITER);
     }
-
     public void remove() {
-        path[--index] = null;
+        // Deletes the last node added to the stringbuilder by deleting from the 2nd last delimiter onwards
+        sb.delete(delimiterIndexes.pop() + 1, sb.length());
```

Comment:
Maybe we can check the stack size and throw a nicer exception than the EmptyStackException, maybe something that mentions an empty contentPath.